main/src/library/RedBlackTree.flix

```diff
@@ -396,4 +398,22 @@ namespace RedBlackTree {
         case _ => tree
     }
 
+    ///
+    /// Extracts a range of key-value pairs from `tree`.
+    ///
+    /// That is, the result is a list of all pairs `f(k, v)` where `p(k)` returns `Equal`.
+    ///
+    pub def queryWith(p: k -> Ordering & e, f: (k, v) -> a, tree: RedBlackTree[k, v]): List[a] & e = {
+        use Core/Cmp.Ordering.{LessThan, Equal, GreaterThan};
+        match tree {
+            case Node(_, a, k, v, b) =>
+                match p(k) {
+                    case LessThan => queryWith(p, f, b)
+                    case Equal => queryWith(p, f, a) ::: f(k, v) :: queryWith(p, f, b)
```

Comment:
Using `:::` is going to be too slow. (Its n^2). I think, in the interest of performance, it would be best to use a mutable array list!

Reply:
Yeah you are probably right. Two questions.
1. Do you still want the final result to be an immutable List?
2. Should I cast such that queryWith is pure or is it acceptable that it is impure?

Reply:
1. Not sure, but I think it should be a list.
2. Its totally legitimate to use interior mutability and then cast to pure :)